conceptnet_sme/semantic_matching.py

```diff
@@ -420,44 +478,225 @@ def forward(self, rels, terms_L, terms_R):
+    def score_edges(
+            self,
+            edges,
+            batch_size=2,
+            convert_logits_to_probas=False,
+            device=None,
+            use_multiple_gpus=False
+    ):
+        """
+        Given a SemanticMatchingModel and an iteratable of edges (in the form of 
+        tuples starting with a relation, a left endpoint, and a right endpoint, 
+        where the endpoints have been run through uri_prefix), return an iterator 
+        yielding the same tuples paired with the model's scores to the edge, 
+        for all of the given edges whose endpoints are present in the model's 
+        term index and whose relation is in the model's relation index.
+        
+        Scores are returned in units of logits, unless conversion to probabilities 
+        is requested by setting the optional paramenter convert_logits_to_probas 
+        to True.  If a device is specified (with the device parameter) then the 
+        model will be moved to that device before scoring takes place; otherwise 
+        the model will be moved to the device given by its device attribute (if not 
+        already there).  After that, if use_multiple_gpus (default False) is 
+        set to True, the model will be parallelized across all available gpu's 
+        (using DataParalleizedModule).  Edges are scored in batches, with a default 
+        size of 2 (which may be changed by setting the batch_size parameter).
+        """
+        # If a device is specified, move the model there before applying it.
+        # Regardless, input for the model will be constructed in batches on the
+        # cpu then moved to the device where the model is, and outputs will be
+        # moved from there to the cpu.
+        if device is None:
+            device = self.device
+        self.to(device)
+
+        # Put the model on all available gpu's, if so requested.
+        parallel_model = DataParallelizedModule(
+            self,
+            device,
+            copy_fn=_model_copier,
+            parallelize=use_multiple_gpus)
+
+        # Put the model(s) in evaluation mode.
+        for model_copy in parallel_model.children:
+            model_copy.eval()
+
+        # We'll batch the edges together before handing them to the model for
+        # evaluation, for efficiency's sake.  As torch tensors are stored row-major,
+        # and the model expects three tensors as input, we make batches of size
+        # 3 x b, then split them by rows to feed them to the model.
+        input_batch = torch.empty(
+            [3, batch_size], dtype=torch.int64, requires_grad=False)
+        input_batch.pin_memory() # speed up any transfer to gpu
+        input_batch_on_device = torch.empty_like(input_batch, device=device)
+        input_batch_edges = []
+
+        position_in_batch = 0
+        for edge in edges:
+            # Translate the string fields of the edge to indices the model knows.
+            rel = edge[0]
```

Comment:
Minor thing: this could be written more simply as `rel, left, right = edge[:3]`.

Reply:
Fixed.